File: 3546. Equal Sum Grid Partition I.cpp
class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) 
    {
        int n = grid.size(), m = grid[0].size();

        long allRowSum = 0, allColSum = 0;
        vector<long> rowSum(n), colSum(m);
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                allRowSum += grid[i][j], allColSum += grid[i][j];
                rowSum[i] += grid[i][j], colSum[j] += grid[i][j];
            }
        }

        long curSum = 0;
        for(auto &sum : rowSum)
        {
            curSum += sum;
            if(curSum + curSum == allRowSum)
                return true;
        }

        curSum = 0;
        for(auto &sum : colSum)
        {
            curSum += sum;
            if(curSum + curSum == allRowSum)
                return true;
        }

        return false;
    }
};

File: 3546. Equal Sum Grid Partition I.py
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:

        total_sum = 0
        for nums in grid:
            total_sum += sum(nums)
        if total_sum % 2:
            return False
        
        cur_sum = 0
        n, m = len(grid), len(grid[0])
        for i in range(n):
            cur_sum += sum(grid[i])
            if cur_sum + cur_sum == total_sum:
                return True

        cur_sum = 0
        for j in range(m):
            for i in range(n):
                cur_sum += grid[i][j]
            if cur_sum + cur_sum == total_sum:
                return True
        return False
        